test_unit/test_sink.cpp

```diff
@@ -40,6 +40,28 @@ using namespace g3;
    EXPECT_TRUE(1 == count->load());
 }
 
+TEST(Sink, OneSinkRemove) {
+using namespace g3;
+   AtomicBoolPtr flag = make_shared < atomic<bool >> (false);
+   AtomicIntPtr count = make_shared < atomic<int >> (0);
+   {
+      auto worker = g3::LogWorker::createLogWorker();
+      auto handle = worker->addSink(std2::make_unique<ScopedSetTrue>(flag, count), &ScopedSetTrue::ReceiveMsg);
+      EXPECT_FALSE(flag->load());
+      EXPECT_TRUE(0 == count->load());
+
+      LogMessagePtr message{std2::make_unique<LogMessage>("test", 0, "test", DEBUG)};
+      message.get()->write().append("this message should trigger an atomic increment at the sink");
+      worker->save(message);
+      worker->removeSink(std::move(handle));
```

Comment:
Regarding the race. Don't feel that you have to follow too closely the test above. It uses RAII to enforce the flush. For your test I think you need to rely on something else.

Example:
1. For the first message you can "cheat" by saving directly to the sink through it's handle
2. Then wait for the returned future, and verify increment of count
3. After this you can remove the sink
4. Then sending a message exactly as you are already doing. No sink should then exist to receive that message
5. Enforce flush of logs through RAII
5. Verify that the last log entry did not increment the count (since the sink was gone).

Doing it this way will remove the race between sending and checking for he counter increment